### PR TITLE
New developer docs:  TripalFieldFormatter, pagers and creating field instances.

### PR DESCRIPTION
**Pull request by spficklin**
# Documentation

Issue #165

## Description

One solution for issue #165  was to provide documentation for how to use the AJAX pager for fields that had too much data to show it all.  But this opened up a can of worms in that there was no documenation for how to setup a formatter class and there was no page for how to create fields and field instances and allow users to add fields via the web interface.  This PR adds all of the documentation.

## Testing?

Just pull the branch, and rebuild the docs.

### CODE REVIEW

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst]: > Therefore, on line 4 of the code above we set the ``$content`` variable to contain this value.

This line is a bit confusing since it doesn't apply to the code block directly above the sentence but the one before that.

```suggestion
You may recall that the ``$items`` array structure is the same as that created by the ``load()`` function described in the :doc:`manual_field_creation` page. Note that each key in the ``value`` array is an accession for a controlled vocabulary term.  These accessions are used to unambiguously describe the value. To display the organism on a page we need the element named ``rdfs:label``.  Thus, we set the ``$content`` variable to contain this value as shown on line 4 of the ``view()`` function above.
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst]: Same concerns with the following sentence:

>Lines 5-6 above use this information to create a clickable link to the organism page.

My suggestion simply specifies that it's the `view()` function code block.

```suggestion
Because our organisms are also published entities we want to link to their respective pages each time an organism is displayed.  Because the ``value`` array has an element named ``entity`` we know that this item is published.  Lines 5-6 of the ``view()`` function shown above use this information to create a clickable link to the organism page.   Finally, the ``$element`` argument is set to provide content of type ``markup``.  This ``$element`` array is a `Drupal renderable array <https://www.drupal.org/docs/7/api/render-arrays/render-arrays-overview>`_.
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst]:
```suggestion
  You should never show the user any data that is outside of ``value`` element.  
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst:76]: Just adding a few more points here that might trip people up.

```suggestion
  - A field with more than one value can have any number of elements in the ``$items`` array.  You should therefore iterate through all of them.
  - For every index in ``$item`` you should create a matching index in ``$element``to display the data found in that ``$item``.
  - If there are no items, then nothing you return will be displayed.
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst:100]: This sentence still leaves me wondering how to set the index properly...

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst]: It could be a bit more clear what the ``$current_page`` variable holds. Is it simple the index (e.g. `3`) of the current page or is it all the items for the current page? An example of how I use this variable to process my items array would be helpful.

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (laceysanderson) [inline docs/dev_guide/custom_field/custom_formatter.rst:180]: This has often tripped me up!

```suggestion

.. warning::

    In order for the ``settingsForm()`` implemented to be available on the "Manage Display" page, you must also implement ``settingsSummary()`` as described below.
```

**Review by laceysanderson — APPROVED** [review]
I built the documentation and it looks as expected :-) 

Great documentation --this is So Needed! There are a few minor suggestions but overall, I approve these docs!

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (spficklin) [reply docs/dev_guide/custom_field/custom_formatter.rst:100]: Unfortunately, this is a problem.   The real answer is that the index has to be a unique number for the page.  But you don't know how many fields there will be on the page that have pagers.  So, the unfortunately reality is you don't know what to set it as.  We need to adjust the code so that this index gets set automatically for the user and they don't have to worry about it.

Comment (spficklin) [reply docs/dev_guide/custom_field/custom_formatter.rst:100]: I added a new issue to address this problem  #783

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (spficklin) [reply docs/dev_guide/custom_field/custom_formatter.rst]: I've adjusted the text to make this more clear.

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (spficklin) [reply docs/dev_guide/custom_field/custom_formatter.rst:100]: So as not to hold up this documentation and to resolve issue #165 I'm going to merge this change. But issue #783 should address this concern for you @laceysanderson

**Comment by spficklin** [comment]
@bradfordcondon and @almasaeed2010 would either of you care to take a look at this before I merge?

File: docs/dev_guide/custom_field/create_instance.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/create_instance.rst]:
```suggestion
 The three TripalField classes simply define how the field will function, but Drupal does not yet know about the field.  The ``hook_bundle_field_info`` function tells Drupal about your field. It must be implemented in a custom Drupal module, and provides an array that tells Drupal about the fields and the classes to use for the field.  Suppose we were creating a field named ``obi__genus`` which displays the Genus for a species and we have a custom module named ``tripal_org2``.  The hook function would be named ``tripal_org2_bundle_field_info()``:
```

typo (module,and).

I also notice we dont bring up the libraries option...

File: docs/dev_guide/custom_field/create_instance.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/create_instance.rst]:
```suggestion
This function receives as its second argument the ``$bundle`` object. This is the bundle that Drupal is requesting new fields for.  For this example we only want to attach the field if the content type is the organism content type.  The format of the returned ``$info`` array should have the field name as the key and an array that follows the instructions provided by Drupal's `field_create_field() <https://api.drupal.org/api/drupal/modules%21field%21field.crud.inc/function/field_create_field/7.x>`_ function. 
```

#### typos
it's
#### suggestions

bundle -> bundle object

File: docs/dev_guide/custom_field/create_instance.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/create_instance.rst:104]: I ALWAYS found this confusing.  Can we please elaborate on the case where base table != chado_table ?

Maybe a ..note directieve after this:

```rst
.. note

   When does the chado_table not match the base_table?  Consider the contact linker field when attached to an organism.  In this case, the base table is the  contact field, and the chado table is the organism_contact table
```
I'm actually not even 100% sure about that since i still find it confusing hahah

File: docs/dev_guide/custom_field/create_instance.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/create_instance.rst]:
```suggestion
Next, we must let Drupal know that our field exists.  We do this by adding an entry to the ``$info`` array of in the ``hook_bundle_field_info()`` function described above.  This lets Drupal know about our field. However, because we are not programmatically creating an instance of the field on a content type, but allowing the user to create them we do not need to implement the ``hook_bundle_instance_info()`` function. Instead, we must implement the ``hook_bundle_create_user_field()``.  This function is called when the user attempts to add our new field to a bundle.  One field that comes with Tripal is the ``chado_linker__prop`` field.  Most Chado base tables have an associated property table (e.g. ``organismprop``, ``featureprop``, ``stockprop``, etc). By default, the ``tripal_chado`` module automatically adds this field to all bundles that have existing properties. It adds a new instance for every property type.  However, new properties can be added to bundle, and the site admin may want to add those properties via the user interface rather. Therefore, this field has the ``$no_ui`` set to TRUE and uses the  ``hook_bundle_create_user_field()`` to create the new field instance for the user.
```

#### typos

let's -> lets
propert
instnace
``$no_ui`` needs double `

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]:
```suggestion
The formatter class is the simplest of all the Tripal field classes.  Here we will again use the **obi__organism** field that comes with the ``tripal_chado`` module.  
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]: maybe explain why
```suggestion
  -  ``$langcode`` is the language. This is used by Drupal to provide translations of data into other spoken languages. By default, Tripal does not use a language, as biological data is generally language agnostic.  Consider for example a gene sequence or a feature coordinate.
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]: explain why?  because webservices right?

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]:
```suggestion
  - For every index in ``$item`` you should create a matching index in ``$element`` to display the data found in that ``$item``.
```
note formatting:

screen shot 2018-12-13 at 10 51 44 am

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst:84]: again i find these sort of directives easier to follow, and to remember, if i know why im being told to do it.

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]:
```suggestion
By default, all links in the pager cause the page to reload.  We do not want the page to reload, rather we only want to update the contents of the field.  The TripalFieldFormatter class provides a function named ``ajaxifyPager`` to convert a pager into an AJAX pager:
```

File: docs/dev_guide/custom_field/custom_formatter.rst
Comment (bradfordcondon) [inline docs/dev_guide/custom_field/custom_formatter.rst]:
```suggestion
An example of the ``settingsSummary()`` function that generates the summary in the image above is as follows:
```

**Review by bradfordcondon — CHANGES_REQUESTED** [review]
looks good.  Thank you very much.  These guides were much needed and I'll get use out of referring to them.  i think you'll find most of my requested changes obvious typo fixes, but one or two are requests for elaboration on the "why" of things

**Comment by spficklin** [comment]
Thanks @bradfordcondon I merged in your suggestions and I updated the docs to respond to your requsted changes.   Thanks for the review!

**Pull request merged**
